File: Edit-distance.java
/*Time Complexity: O(M*N)

Space Complexity: O(M*N + M*N)

Did this code successfully run on Leetcode : Yes

Approach: Memoization - Exploring all possiblies and storing repeated subprobs in dp array.

Prob: 72. Edit Distance
*/

import java.util.Arrays;

class Solution {
    int helper(String word1, String word2, int i,int j, int [][] dp){
        if(i < 0) 
            return j+1;
        if(j < 0) 
            return i+1;
        if(dp[i][j] != -1){
            return dp[i][j];
        }
        if(word1.charAt(i) == word2.charAt(j)){
            return dp[i][j] = 0 + helper(word1,word2,i-1,j-1,dp);
        }
        else{
            return dp[i][j] = 1 + Math.min(helper(word1,word2,i-1,j,dp),
            Math.min(helper(word1,word2,i,j-1,dp),helper(word1,word2,i-1,j-1,dp)));
        }
    }
    public int minDistance(String word1, String word2) {
        int n = word1.length();
        int m = word2.length();
        int [][] dp = new int[n][m];
        for(int []rows : dp){
            Arrays.fill(rows,-1);
        }
        return helper(word1,word2,n-1,m-1,dp);
    }
}

File: EditDistance.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace DynamicProgramming
{
    public class EditDistance
    {
        /*
         * T.C: O(M * N) where M is length of word1 and N is word of length 2
         * S.C: O(M *N) as dp array is of length m *n
         */
        public int MinDistance(string word1, string word2)
        {
            if (word1.Equals(word2)) return 0;

            int n = word1.Length;
            int m = word2.Length;

            int[,] dp = new int[m + 1, n + 1];


            if (m > n) MinDistance(word2, word1);

            for (int i = 0; i < n + 1; i++)
            {
                dp[0, i] = i;
            }

            for (int i = 0; i < m + 1; i++)
            {
                dp[i, 0] = i;
            }

            for (int i = 1; i < m + 1; i++)
            {
                for (int j = 1; j < n + 1; j++)
                {
                    if (word2[i - 1] == word1[j - 1])
                    {
                        dp[i, j] = dp[i - 1, j - 1];
                    }
                    else
                    {
                        dp[i, j] = Math.Min(dp[i - 1, j], Math.Min(dp[i - 1, j - 1], dp[i, j - 1])) + 1;
                    }
                }
            }

            return dp[m, n];
        }


        /*
         * T.C: O(m*n)
         * S.C: O(n)
         * 
         */

        public int MinDistance1(string word1, string word2)
        {
            if (word1.Equals(word2)) return 0;

            int n = word1.Length;
            int m = word2.Length;

            int[] dp = new int[n + 1];


            if (m > n) MinDistance1(word2, word1);

            for (int i = 0; i < n + 1; i++)
            {
                dp[i] = i;
            }


            for (int i = 1; i < m + 1; i++)
            {
                int temp = dp[0];
                dp[0] = i;
                for (int j = 1; j < n + 1; j++)
                {
                    int prev = dp[j];
                    if (word2[i - 1] == word1[j - 1])
                    {
                        dp[j] = temp;
                    }
                    else
                    {
                        dp[j] = Math.Min(dp[j - 1], Math.Min(temp, dp[j])) + 1;
                    }
                    temp = prev;
                }
            }

            return dp[n];
        }
    }
}
